File: app/src/main/java/cf/movie/slmovie/main/detail/ui/DetailActivity.java
package cf.movie.slmovie.main.detail.ui;

import android.content.Intent;
import android.graphics.Color;
import android.support.design.widget.CoordinatorLayout;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.Toolbar;
import android.view.KeyEvent;
import android.view.MenuItem;
import android.widget.Toast;

import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactInstanceManagerBuilder;
import com.facebook.react.ReactRootView;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import com.facebook.react.shell.MainReactPackage;

import cf.movie.slmovie.R;
import cf.movie.slmovie.base.BaseActivity;
import cf.movie.slmovie.main.detail.presenter.DetailPresenter;
import cf.movie.slmovie.main.detail.rn.DetailReactPackage;
import cf.movie.slmovie.utils.rnUtils.checkVersion.CheckVersionReactPackage;
import cf.movie.slmovie.utils.rnUtils.toastDialog.ToastDialogReactPackage;
import cf.movie.slmovie.utils.rnUtils.download.DownloadReactPackage;
import cf.movie.slmovie.utils.rnUtils.snackbar.SnackbarReactPackage;

/**
 * Created by 包俊 on 2017/8/5.
 */

public class DetailActivity extends BaseActivity implements DefaultHardwareBackBtnHandler {
    private ReactRootView mReactRootView;
    private ReactInstanceManager mReactInstanceManager;
    private Toolbar toolbar;
    private CoordinatorLayout container;
    private SwipeRefreshLayout swipeLayout;
    private String address;
    private DetailPresenter presenter;

    @Override
    protected int getContentLayout() {
        return R.layout.activity_detail;
    }

    @Override
    protected void initGui() {
        mReactRootView = (ReactRootView) findViewById(R.id.react);
        toolbar = (Toolbar) findViewById(R.id.toolbar);
        container = (CoordinatorLayout) findViewById(R.id.container);
        swipeLayout = (SwipeRefreshLayout) findViewById(R.id.swipeLayout);
    }

    @Override
    protected void initData() {
        Intent intent = getIntent();
        address = intent.getStringExtra("address");
        String name = intent.getStringExtra("name");
        setSupportActionBar(toolbar);
        getSupportActionBar().setTitle(name);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        swipeLayout.setColorSchemeColors(Color.BLUE,
                Color.GREEN,
                Color.YELLOW,
                Color.RED);
        swipeLayout.setProgressBackgroundColorSchemeColor(Color.WHITE);
        swipeLayout.setSize(SwipeRefreshLayout.DEFAULT);
        swipeLayout.setEnabled(false);
        swipeLayout.setRefreshing(true);
        presenter = new DetailPresenter(this, iDetailView);
        if (presenter.checkVersion()) {
            mReactInstanceManager = presenter.initReact().build();
            mReactRootView.startReactApplication(mReactInstanceManager, "DetailActivity", null);
        } else {
            Toast.makeText(this, "失败", Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void initAction() {

    }

    private IDetailView iDetailView = new IDetailView() {
        /**
         * swipeLayout刷新状态
         *
         * @param fresh
         */
        @Override
        public void refresh(final boolean fresh) {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    swipeLayout.setRefreshing(fresh);
                }
            });
        }

        @Override
        public void setReactPackage(ReactInstanceManagerBuilder builder) {
            builder.addPackage(new MainReactPackage())
                    .addPackage(new DetailReactPackage(address, iDetailView))
                    .addPackage(new SnackbarReactPackage(container))
                    .addPackage(new DownloadReactPackage())
                    .addPackage(new ToastDialogReactPackage(DetailActivity.this))
                    .addPackage(new CheckVersionReactPackage(iDetailView));
        }

        @Override
        public void reCreateReactNative() {
            runOnUiThread(new Runnable() {
                @Override
                public void run() {
                    mReactInstanceManager.recreateReactContextInBackground();
                }
            });
        }
    };

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void invokeDefaultOnBackPressed() {
        super.onBackPressed();
    }

    @Override
    protected void onPause() {
        super.onPause();

        if (mReactInstanceManager != null) {
            mReactInstanceManager.onHostPause(this);
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if (mReactInstanceManager != null) {
            mReactInstanceManager.onHostResume(this, this);
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (mReactInstanceManager != null) {
            mReactInstanceManager.onHostDestroy();
        }
    }

    @Override
    public void onBackPressed() {
        if (mReactInstanceManager != null) {
            mReactInstanceManager.onBackPressed();
        } else {
            super.onBackPressed();
        }
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_MENU && mReactInstanceManager != null) {
            mReactInstanceManager.showDevOptionsDialog();
            return true;
        }
        return super.onKeyUp(keyCode, event);
    }

    private float up, down;

}

File: app/src/main/java/cf/movie/slmovie/main/detail/ui/DetailActivity.kt
package cf.movie.slmovie.main.detail.ui

import android.content.Intent
import android.graphics.Color
import android.support.design.widget.CoordinatorLayout
import android.support.v4.widget.SwipeRefreshLayout
import android.support.v7.widget.Toolbar
import android.view.KeyEvent
import android.view.MenuItem
import android.widget.Toast

import com.facebook.react.ReactInstanceManager
import com.facebook.react.ReactInstanceManagerBuilder
import com.facebook.react.ReactRootView
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler
import com.facebook.react.shell.MainReactPackage

import cf.movie.slmovie.R
import cf.movie.slmovie.base.BaseActivity
import cf.movie.slmovie.main.detail.presenter.DetailPresenter
import cf.movie.slmovie.main.detail.rn.DetailReactPackage
import cf.movie.slmovie.utils.rnUtils.checkVersion.CheckVersionReactPackage
import cf.movie.slmovie.utils.rnUtils.toastDialog.ToastDialogReactPackage
import cf.movie.slmovie.utils.rnUtils.download.DownloadReactPackage
import cf.movie.slmovie.utils.rnUtils.snackbar.SnackbarReactPackage

/**
 * Created by 包俊 on 2017/8/5.
 */

class DetailActivity : BaseActivity(), DefaultHardwareBackBtnHandler {
    private var mReactRootView: ReactRootView? = null
    private var mReactInstanceManager: ReactInstanceManager? = null
    private var toolbar: Toolbar? = null
    private var container: CoordinatorLayout? = null
    private var swipeLayout: SwipeRefreshLayout? = null
    private var address: String? = null
    private var presenter: DetailPresenter? = null

    override val contentLayout: Int
        get() = R.layout.activity_detail

    private val iDetailView = object : IDetailView {
        /**
         * swipeLayout刷新状态
         *
         * @param fresh
         */
        override fun refresh(fresh: Boolean) {
            runOnUiThread { swipeLayout!!.isRefreshing = fresh }
        }

        override fun setReactPackage(builder: ReactInstanceManagerBuilder) {
            builder.addPackage(MainReactPackage())
                    .addPackage(DetailReactPackage(address!!, this))
                    .addPackage(SnackbarReactPackage(container!!))
                    .addPackage(DownloadReactPackage())
                    .addPackage(ToastDialogReactPackage(this@DetailActivity))
                    .addPackage(CheckVersionReactPackage(this))
        }

        override fun reCreateReactNative() {
            runOnUiThread { mReactInstanceManager!!.recreateReactContextInBackground() }
        }
    }

    private val up: Float = 0.toFloat()
    private val down: Float = 0.toFloat()

    override fun initGui() {
        mReactRootView = findViewById(R.id.react) as ReactRootView
        toolbar = findViewById(R.id.toolbar) as Toolbar
        container = findViewById(R.id.container) as CoordinatorLayout
        swipeLayout = findViewById(R.id.swipeLayout) as SwipeRefreshLayout
    }

    override fun initData() {
        val intent = intent
        address = intent.getStringExtra("address")
        val name = intent.getStringExtra("name")
        setSupportActionBar(toolbar)
        supportActionBar!!.title = name
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        swipeLayout!!.setColorSchemeColors(Color.BLUE,
                Color.GREEN,
                Color.YELLOW,
                Color.RED)
        swipeLayout!!.setProgressBackgroundColorSchemeColor(Color.WHITE)
        swipeLayout!!.setSize(SwipeRefreshLayout.DEFAULT)
        swipeLayout!!.isEnabled = false
        swipeLayout!!.isRefreshing = true
        presenter = DetailPresenter(this, iDetailView)
        if (presenter!!.checkVersion()) {
            mReactInstanceManager = presenter!!.initReact().build()
            mReactRootView!!.startReactApplication(mReactInstanceManager, "DetailActivity", null)
        } else {
            Toast.makeText(this, "失败", Toast.LENGTH_LONG).show()
        }
    }

    override fun initAction() {

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    override fun invokeDefaultOnBackPressed() {
        super.onBackPressed()
    }

    override fun onPause() {
        super.onPause()

        if (mReactInstanceManager != null) {
            mReactInstanceManager!!.onHostPause(this)
        }
    }

    override fun onResume() {
        super.onResume()

        if (mReactInstanceManager != null) {
            mReactInstanceManager!!.onHostResume(this, this)
        }
    }

    override fun onDestroy() {
        super.onDestroy()

        if (mReactInstanceManager != null) {
            mReactInstanceManager!!.onHostDestroy()
        }
    }

    override fun onBackPressed() {
        if (mReactInstanceManager != null) {
            mReactInstanceManager!!.onBackPressed()
        } else {
            super.onBackPressed()
        }
    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_MENU && mReactInstanceManager != null) {
            mReactInstanceManager!!.showDevOptionsDialog()
            return true
        }
        return super.onKeyUp(keyCode, event)
    }

}
